File: brent.py
import math
import sys
from collections.abc import Callable

macheps = sys.float_info.epsilon

def zero(
    f: Callable[[float], float],
    a: float,
    b: float,
    t: float,
    eps: float | None = None
) -> float:
    """
    Brent's method for root-finding, a direct translation of the ALGOL 60
    procedure from Richard P. Brent's book "Algorithms for Minimization Without
    Derivatives" (Prentice-Hall, 1973). The original description from Brent's
    book is reproduced below.

    Procedure `zero` returns a zero `x` of the function `f` in the interval
    [a, b], to within a tolerance `6*eps*abs(x) + 2*t`, where `eps` is the
    relative machine precision and `t` is a positive tolerance. The procedure
    assumes that `f(a)` and `f(b)` have different signs.

    Parameters
    ----------
    f: Callable
        Function for which to find a root. Should take a single float
        argument and return a float value.
    a, b: float
        Initial bracket: two values on either side of the root, for which
        f(a)*f(b) < 0.
    t: float
        Absolute tolerance for the result. Should be positive.
    eps: float
        Relative tolerance. If `None` (the default), machine epsilon is used.

    Returns
    -------
    x: float
        Estimated value of the root.
    """
    if eps is None:
        eps = macheps

    fa = f(a)
    fb = f(b)

    c, fc = a, fa
    d = b - a
    e = d

    while True:
        if abs(fc) < abs(fb):
            a, fa = b, fb
            b, fb = c, fc
            c, fc = a, fa

        tol = 2*eps*abs(b) + t
        m = 0.5*(c - b)

        if abs(m) <= tol or fb == 0:
            break

        # See if a bisection is forced
        if abs(e) < tol and abs(fa) <= abs(fb):
            e = m
            d = e
        else:
            s = fb/fa
            if a == c:
                # Linear interpolation
                p = 2*m*s
                q = 1 - s
            else:
                # Inverse quadratic interpolation
                q = fa/fc
                r = fb/fc
                p = s*(2*m*q*(q - r) - (b - a)*(r - 1))
                q = (q - 1)*(r - 1)*(s - 1)
            if p > 0:
                q = -q
            else:
                p = -p
            s = e
            e = d
            if 2*p < 3*m*q - abs(tol*q) and p < abs(0.5*s*q):
                d = p/q
            else:
                e = m
                d = e
        a, fa = b, fb
        b += d if abs(d) > tol else (tol if m > 0 else -tol)
        fb = f(b)

        if (fb > 0) == (fc > 0):
            c, fc = a, fa
            d = b - a
            e = d

    return b

def localmin(
    f: Callable[[float], float],
    a: float,
    b: float,
    x: float,
    t: float,
    eps: float | None = None,
) -> float:
    """
    Brent's method for minimization, a direct translation of the ALGOL 60
    procedure from Richard P. Brent's book "Algorithms for Minimization Without
    Derivatives" (Prentice-Hall, 1973). The original description from Brent's
    book is reproduced below.

    If the function `f` is defined on the interval (a, b), then `localmin`
    finds an approximation `x` to the point at which `f` attains its maximum
    (or the appropriate limit point) and returns the value of `x`.
    `t` and `eps` define a tolerance `tol = eps*abs(x) + t`, and `f` is never
    evaluated at two points closer together than `tol`. If `f` is δ-unimodal
    for some δ < `tol`, then `x` approximates the global minimum of `f` with
    an error less than `3*tol`. If `f` is not δ-unimodal on (a, b), then `x`
    may approximate a local, but non-global, minimum. `eps` should be no
    smaller than `2*macheps`, and preferably not much less than `sqrt(macheps)`,
    where `macheps` is the relative machine precision. `t` should be positive.

    The method used is a combination of golden section search and successive
    parabolic interpolation. Convergence is never much slower than for a
    Fibonacci search. If `f` has a continuous second derivative which is
    positive at the minimum (not at `a` or `b`) then, ignoring rounding errors,
    convergence is superlinear, and usually the order is at least 1.3247...

    Parameters
    ----------
    f: Callable
        Function for which to find a root. Should take a single float
        argument and return a float value.
    a, b: float
        Initial bracket: two values on either side of the minimum.
    x: float
        Initial estimate of the minimum. Should satisfy f(x) < f(a) and
        f(x) < f(b).
    t: float
        Absolute tolerance for the result. Should be positive.
    eps: float
        Relative tolerance. If `None`, a default value is calculated based as
        the square root of the machine precision.

    Returns
    -------
    x: float
        Estimated value of the argument at which `f` attains its minimum.
    """
    if eps is None:
        eps = math.sqrt(macheps)

    c = (3 - math.sqrt(5))/2
    x = a + c*(b - a)
    e = 0
    fx = f(x)
    w, fw = x, fx
    v, fv = w, fw

    while True:
        m = 0.5*(a + b)
        tol = eps*abs(x) + t
        t2 = 2*tol

        if abs(x - m) <= t2 - 0.5*(b - a):
            break

        p = 0
        q = 0
        r = 0
        if abs(e) > tol:
            # Fit parabola
            r = (x - w)*(fx - fv)
            q = (x - v)*(fx - fw)
            p = (x - v)*q - (x - w)*r
            q = 2*(q - r)
            if q > 0:
                p = -p
            else:
                q = -q
            r = e
            e = d
        if abs(p) < abs(0.5*q*r) and p < q*(a - x) and p < q*(b - x):
            # A "parabolic interpolation" step
            d = p/q
            u = x + d
            # f must not be evaluated too close to a or b
            if u - a < t2 and b - u < t2:
                d = tol if x < m else -tol
        else:
            # A "golden section" step
            e = (b if x < m else a) - x
            d = c*e

        # f must not be evaluated too close to x
        u = x + (d if abs(d) >= tol else (tol if d > 0 else -tol))
        fu = f(u)

        # Update a, b, v, w, and x
        if fu <= fx:
            if u < x:
                b = x
            else:
                a = x
            v, fv = w, fw
            w, fw = x, fx
            x, fx = u, fu
        else:
            if u < x:
                a = u
            else:
                b = u
            if fu <= fw or w == x:
                v, fv = w, fw
                w, fw = u, fu
            elif fu <= fv or v == x or v == w:
                v, fv = u, fu

    return x

File: brent.f
C     A FORTRAN TRANSLATION OF THE ALGOL PROCEDURE ZERO.
C     SEE PROCEDURE ZERO, SECTION 4.6, FOR COMMENTS ETC.
      REAL FUNCTION ZERO (A, B, MACHEP, T, F)
      REAL A,B,MACHEP,T,F,SA,SB,C,D,E,FA,FB,FC,TOL,M,P,Q,R,S
      SA = A
      SB = B
      FA = F(SA)
      FB = F(SB)
   10 C = SA
      FC = FA
      E = SB - SA
      D = E
   20 IF (ABS(FC).GE.ABS(FB)) GO TO 30
      SA = SB
      SB = C
      C = SA
      FA = FB
      FB = FC
      FC = FA
   30 TOL = 2.0*MACHEP*ABS(SB) + T
      M = 0.5*(C - SB)
      IF ((ABS(M).LE.TOL).OR.(FB.EQ.0.0)) GO TO 140
      IF ((ABS(E).GE.TOL).AND.(ABS(FA).GT.ABS(FB))) GO TO 40
      E = M
      D = E
      GO TO 100
   40 S = FB/FA
      IF (SA.NE.C) GO TO 50
      P = 2.0*M*S
      Q = 1.0 - S
      GO TO 60
   50 Q = FA/FC
      R = FB/FC
      P = S*(2.0*M*Q*(Q - R) - (SB - SA)*(R - 1.0))
      Q = (Q - 1.0)*(R - 1.0)*(S - 1.0)
   60 IF (P.LE.0.0) GO TO 70
      Q = -Q
      GO TO 80
   70 P = -P
   80 S = E
      E = D
      IF ((2.0*P.GE.3.0*M*Q-ABS(TOL*Q)).OR.(P.GE.ABS(0.5*S*Q))) GO TO 90
      D = P/Q
      GO TO 100
   90 E = M
      D = E
  100 SA = SB
      FA = FB
      IF (ABS(D).LE.TOL) GO TO 110
      SB = SB + D
      GO TO 130
  110 IF (M.LE.0.0) GO TO 120
      SB = SB + TOL
      GO TO 130
  120 SB = SB - TOL
  130 FB = F(SB)
      IF ((FB.GT.0.0).AND.(FC.GT.0.0)) GO TO 10
      IF ((FB.LE.0.0).AND.(FC.LE.0.0)) GO TO 10
      GO TO 20
  140 ZERO = SB
      RETURN
      END
C     A FORTRAN TRANSLATION OF THE ALGOL PROCEDURE LOCALMIN.
C     SEE PROCEDURE LOCALMIN, SECTION 5.8, FOR COMMENTS ETC.
      REAL FUNCTION LOCALM (A, B, EPS, T, F, X)
      REAL A,B,EPS,T,F,X,SA,SB,D,E,M,P,Q,R,TOL,T2,U,V,W,FU,FV,FW,FX
      SA = A
      SB = B
      X = SA + 0.381966*(SB - SA)
      W = X
      V = W
      E = 0.0
      FX = F(X)
      FW = FX
      FV = FW
   10 M = 0.5*(SA + SB)
      TOL = EPS*ABS(X) + T
      T2 = 2.0*TOL
      IF (ABS(X-M).LE.T2-0.5*(SB-SA)) GO TO 190
      R = 0.0
      Q = R
      P = Q
      IF (ABS(E).LE.TOL) GO TO 40
      R = (X - W)*(FX - FV)
      Q = (X - V)*(FX - FW)
      P = (X - V)*Q - (X - W)*R
      Q = 2.0*(Q - R)
      IF (Q.LE.0.0) GO TO 20
      P = -P
      GO TO 30
   20 Q = -Q
   30 R = E
      E = D
   40 IF (ABS(P).GE.ABS(0.5*Q*R)) GO TO 60
      IF ((P.LE.Q*(SA-X)).OR.(P.GE.Q*(SB-X))) GO TO 60
      D = P/Q
      U = X + D
      IF ((U-SA.GE.T2).AND.(SB-U.GE.T2)) GO TO 90
      IF (X.GE.M) GO TO 50
      D = TOL
      GO TO 90
   50 D = -TOL
      GO TO 90
   60 IF (X.GE.M) GO TO 70
      E = SB - X
      GO TO 80
   70 E = SA - X
   80 D = 0.381966*E
   90 IF (ABS(D).LT.TOL) GO TO 100
      U = X + D
      GO TO 120
  100 IF (D.LE.0.0) GO TO 110
      U = X + TOL
      GO TO 120
  110 U = X - TOL
  120 FU = F(U)
      IF (FU.GT.FX) GO TO 150
      IF (U.GE.X) GO TO 130
      SB = X
      GO TO 140
  130 SA = X
  140 V = W
      FV = FW
      W = X
      FW = FX
      X = U
      FX = FU
      GO TO 10
  150 IF (U.GE.X) GO TO 160
      SA = U
      GO TO 170
  160 SB = U
  170 IF ((FU.GT.FW).AND.(W.NE.X)) GO TO 180
      V = W
      FV = FW
      W = U
      FW = FU
      GO TO 10
  180 IF ((FU.GT.FV).AND.(V.NE.X).AND.(V.NE.W)) GO TO 10
      V = U
      FV = FU
      GO TO 10
  190 LOCALM = FX
      RETURN
      END
